File: tag_detector/newDetector.py
import cv2
import numpy as np
import matplotlib.pyplot as plt

fig = plt.figure()
ax = fig.add_subplot(projection='3d')

ax.set_xlim([-1000, 1000])  # testing with -1 to 1 meter range for each axis
ax.set_ylim([-1000, 1000])
ax.set_zlim([-1000, 1000])

point, = ax.plot([], [], [], 'ro', markersize=8)
orig, = ax.plot([], [], [], 'bo', markersize=8)

def update_3d_plot(tvec):
    if tvec is not None:
        x = tvec[2].item()
        y = tvec[0].item()
        z = tvec[1].item()
        point.set_data([x], [y])
        point.set_3d_properties([z])
        orig.set_data([0], [0])
        orig.set_3d_properties([0])
        plt.draw()
        plt.pause(0.01)

def detect_target(frame):
    hsv = cv2.cvtColor(frame, cv2.COLOR_BGR2HSV)
    
    # Define red color range in HSV
    lower_red1 = np.array([0, 120, 70])
    upper_red1 = np.array([10, 255, 255])
    lower_red2 = np.array([170, 120, 70])
    upper_red2 = np.array([180, 255, 255])

    lowerBlue = np.array([90,80,160])
    upperBlue = np.array([120,255,255])
    
    # Create masks to detect red color
    mask1 = cv2.inRange(hsv, lower_red1, upper_red1)
    mask2 = cv2.inRange(hsv, lower_red2, upper_red2)
    maskRed = mask1 + mask2

    maskBlue = cv2.inRange(hsv, lowerBlue, upperBlue)
    
    # Apply mask to extract red regions
    red_regions = cv2.bitwise_and(frame, frame, mask=maskRed)
    blue_regions = cv2.bitwise_and(frame, frame, mask = maskBlue)

    grayRed = cv2.cvtColor(red_regions, cv2.COLOR_BGR2GRAY)
    grayBlue = cv2.cvtColor(blue_regions, cv2.COLOR_BGR2GRAY)
    # cv2.imshow("blue", maskBlue)
    
    edgesRed = cv2.Canny(grayRed, 50, 150)
    edgesB = cv2.Canny(grayBlue, 50, 150)
    # cv2.imshow("edges", edgesB)
    contoursRed, _ = cv2.findContours(edgesRed, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    contoursB, _ = cv2.findContours(edgesB, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    contoursRed = sorted(contoursRed, key=cv2.contourArea, reverse=True)
    contoursB = sorted(contoursB, key=cv2.contourArea, reverse=True)
    
    approx_cornersRed = None
    approx_cornersB = None
    min_area_threshold = 500  # Ignore small detections

    # Do edge filtering for both Red and Blue contours, return all contours that fit criteria
    contour_listR = list()
    for contourRed in contoursRed:
        epsilon = 0.02 * cv2.arcLength(contourRed, True)
        approxRed = cv2.approxPolyDP(contourRed, epsilon, True)
        
        if len(approxRed) == 4:
            areaRed = cv2.contourArea(approxRed)
            if areaRed < min_area_threshold:
                continue  # Ignore small contours
            
            # Convexity check
            if not cv2.isContourConvex(approxRed):
                continue
            
            approx_cornersRed = approxRed
            if approx_cornersRed is not None:
                cv2.drawContours(frame, [approx_cornersRed], -1, (0, 255, 0), 3)
                contour_listR.append(approx_cornersRed.reshape(4, 1, 2))
            
    contour_listB = list()
    for contourB in contoursB:
        epsilon = 0.02 * cv2.arcLength(contourB, True)
        approxB = cv2.approxPolyDP(contourB, epsilon, True)
        
        if len(approxB) == 4:
            areaB = cv2.contourArea(approxB)
            if areaB < min_area_threshold:
                continue  # Ignore small contours
            
            # Convexity check
            if not cv2.isContourConvex(approxB):
                continue
            
            approx_cornersB = approxB
            if approx_cornersB is not None:
                cv2.drawContours(frame, [approx_cornersB], -1, (0, 255, 0), 3)
                contour_listB.append(approx_cornersB.reshape(4, 1, 2))
        
    return frame, contour_listR, contour_listB

def order(pts):
    rect = np.zeros((4, 2), dtype="float32")

    s = pts.sum(axis=1)
    # print(np.argmax(s))
    rect[0] = pts[np.argmin(s)]
    rect[2] = pts[np.argmax(s)]

    diff = np.diff(pts, axis=1)
    # print(np.argmax(diff))
    rect[1] = pts[np.argmin(diff)]
    rect[3] = pts[np.argmax(diff)]

    # return the ordered coordinates
    return rect

def determineLetter(marker):
    ret, img_bw = cv2.threshold(marker, 200, 255, cv2.THRESH_BINARY)
    white = 255
    if img_bw[5, 5] == white:
        return False
    cropped_img = img_bw[25:150, 25:150]

    # [y, x]: [0, 0] = top left pixel
    # Each pixel seems to be like 25 
    if cropped_img[12, 37] != white and cropped_img[12, 62] == white:    # Center of pixel at [0, 1]; not white if letter "A"
        cv2.imshow("Cropped", cropped_img)
        return "A"
    elif cropped_img[12, 12] != white and cropped_img[12, 37] == white:  # Center of pixel at [0, 0]; not white if letter "C"
        cv2.imshow("Cropped", cropped_img)
        return "C"
    elif cropped_img[12, 112] == white and cropped_img[37, 37] != white: # Center of pixel at [0, 4]; white if letter "E"
        cv2.imshow("Cropped", cropped_img)
        return "E"
    elif cropped_img[62, 112] == white and cropped_img[12, 112] != white: # Center of pixel at [2, 4]; white if letter "D"
        cv2.imshow("Cropped", cropped_img)
        return "D"
    elif cropped_img[37, 112] == white and cropped_img[62, 122] != white:
        cv2.imshow("Cropped", cropped_img)
        return "B"                      # Define B as having [1, 4] and [2, 3] as white
    
    return None

fid_size = 53  # centimeters
# TODO fix calibration to setup cameraMatrix
cameraMatrix = np.array([[1.25649815e+03, 0.0, 7.12996774e+02],
                         [0.0, 1.25820533e+03, 4.69551858e+02], 
                         [0.0, 0.0, 1.0]], dtype=np.float32)
distCoeffs = np.array([[-3.72271817e-03, 5.33786890e-01, -4.99625728e-04, -1.65101232e-03,-1.78505927e+00]], dtype=np.float32)
# TODO testing program with hardcoded matrix since calibration session is not working
# TODO when game ready, dont forget to comment this out

def findTranslationAndRotation(image_pts):
    object_points = np.array([[-fid_size / 2.0, fid_size / 2.0, 0.0],
                              [fid_size / 2.0, fid_size / 2.0, 0.0],
                              [fid_size / 2.0, -fid_size / 2.0, 0.0],
                              [-fid_size / 2.0, -fid_size / 2.0, 0.0]], dtype=np.float32)
    #TODO: make calibration dynamic somehow
    _, rvec, tvec = cv2.solvePnP(object_points, np.array(image_pts, dtype=np.float32), 
                                 cameraMatrix, distCoeffs) #removed IPPE_SQUARE flag
    if rvec is not None:
        Rt = np.matrix(cv2.Rodrigues(rvec)[0])
        # TODO: remove all the yaw, pitch roll stuff at some point
        yaw, pitch, roll = rotToEul(Rt)
        R = Rt.T
        pose = -R * np.matrix(tvec)
        return tvec, (yaw, pitch, roll) #returning raw tvec rather than pose for testing.
    
    return "ERROR: TVEC nfound", "ERROR: rvec nfound"

# This function is kinda useless cuz we don't need it - TODO: remove it later
def rotToEul(R):
    sy = np.sqrt(R[0, 0] ** 2 + R[1, 0] ** 2)

    singular = sy < 1e-6
    if not singular:
        yaw = np.arctan2(R[2, 1], R[2, 2])  # Rotation around Z-axis
        pitch = np.arctan2(-R[2, 0], sy)    # Rotation around Y-axis
        roll = np.arctan2(R[1, 0], R[0, 0]) # Rotation around X-axis
    else:
        yaw = np.arctan2(-R[1, 2], R[1, 1])
        pitch = np.arctan2(-R[2, 0], sy)
        roll = 0

    # Convert radians to degrees
    return np.degrees([yaw, pitch, roll])

def main():
    cap = cv2.VideoCapture(0)
    
    if not cap.isOpened():
        print("Error: Could not open webcam.")
        return
    
    dim = 175
    p1 = np.array([
        [0, 0],
        [dim - 1, 0],
        [dim - 1, dim - 1],
        [0, dim - 1]], dtype="float32")

    while True:
        ret, frame = cap.read()
        if not ret:
            break
        
        frame, contoursR, contoursB = detect_target(frame)
        if len(contoursR):
            for cornersR in contoursR:
                c_rezR = order(cornersR[:, 0])
                hR, _ = cv2.findHomography(c_rezR, p1, cv2.RANSAC, 2)
                markerLetterR = None
                if hR is not None:
                    tagR = cv2.warpPerspective(frame, hR, (175, 175))
                    markerLetterR = determineLetter(cv2.cvtColor(tagR, cv2.COLOR_BGR2GRAY))
                    if markerLetterR:
                        tvecR, anglesR = findTranslationAndRotation(c_rezR)
                        print("Red Tag: " + markerLetterR)
                        print("CAMERA Position x=" + str(tvecR[2]) + " y=" + str(tvecR[0]) + " z=" + str(tvecR[1]))
                        print("Camera tilt -------   yaw=" + str(anglesR[0]) + " pitch=" + str(anglesR[1]) + " roll=" + str(anglesR[2]))

                        update_3d_plot(tvecR)

        if len(contoursB):
            for cornersB in contoursB:
                c_rezB = order(cornersB[:, 0])
                hB, _ = cv2.findHomography(c_rezB, p1, cv2.RANSAC, 2)
                markerLetterB = None
                if hB is not None:
                    tagB = cv2.warpPerspective(frame, hB, (175, 175))
                    markerLetterB = determineLetter(cv2.cvtColor(tagB, cv2.COLOR_BGR2GRAY))
                    if markerLetterB:
                        tvecB, anglesB = findTranslationAndRotation(c_rezB)
                        print("Blue Tag: " + markerLetterB)
                        print("CAMERA Position x=" + str(tvecB[2]) + " y=" + str(tvecB[0]) + " z=" + str(tvecB[1]))
                        print("Camera tilt -------   yaw=" + str(anglesB[0]) + " pitch=" + str(anglesB[1]) + " roll=" + str(anglesB[2]))
        
        cv2.imshow("Webcam Detection", frame)
        
        if cv2.waitKey(1) & 0xFF == ord('q'):
            break
    
    cap.release()
    cv2.destroyAllWindows()

if __name__ == "__main__":
    main()

File: tag_detector/newDetector.cpp
// tag_localization_dual.cpp
// Build (Ubuntu):
//   g++ -std=c++17 tag_localization_dual.cpp -o tag_loc `pkg-config --cflags --libs opencv4`
// Run:
//   ./tag_loc
#include <opencv2/opencv.hpp>
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <algorithm>
#include <cmath>
#include <iomanip>

using namespace cv;
using std::cout;
using std::endl;

// ----------------- Config / constants -----------------
static const float fid_size_mm = 100.0f; // mm
// Camera intrinsics (hardcoded like your Python)
static const Mat cameraMatrix = (Mat_<float>(3,3) <<
    1.18666112e+03f, 0.f, 7.22383627e+02f,
    0.f, 1.19064020e+03f, 4.94566994e+02f,
    0.f, 0.f, 1.f);
// Zero distortion for testing (like your file)
static const Mat distCoeffs = Mat::zeros(1, 5, CV_32F);

// Field tag poses (x, y, directionDeg)
static const std::map<std::string, cv::Vec3f> TAG_POSES = {
    {"A_RD",{  500.f, 7999.f,  90.f}},
    {"B_RD",{    0.f,  500.f, 180.f}},
    {"C_RD",{ 2999.f, 3500.f,   0.f}},
    {"D_RD",{ 3171.f, 5500.f, 180.f}},
    {"E_RD",{ 5350.f, 7999.f,  90.f}},
    {"A_BL",{11500.f,    0.f, 270.f}},
    {"B_BL",{12000.f, 7500.f,   0.f}},
    {"C_BL",{ 9001.f, 4500.f, 180.f}},
    {"D_BL",{ 8829.3f,2500.f,   0.f}},
    {"E_BL",{ 6750.f,    0.f, 270.f}},
};

// ----------------- Helpers -----------------
static std::vector<Point2f> order_corners(const std::vector<Point2f>& pts) {
    // Same logic as Python 'order': uses sum and diff to assign TL, TR, BR, BL
    CV_Assert(pts.size() == 4);
    std::vector<Point2f> rect(4);
    std::vector<float> sumv, diffv;
    sumv.reserve(4); diffv.reserve(4);
    for (const auto& p : pts) {
        sumv.push_back(p.x + p.y);
        diffv.push_back(p.y - p.x);
    }
    // TL: smallest sum, BR: largest sum
    int i_tl = int(std::distance(sumv.begin(), std::min_element(sumv.begin(), sumv.end())));
    int i_br = int(std::distance(sumv.begin(), std::max_element(sumv.begin(), sumv.end())));
    // TR: smallest diff, BL: largest diff
    int i_tr = int(std::distance(diffv.begin(), std::min_element(diffv.begin(), diffv.end())));
    int i_bl = int(std::distance(diffv.begin(), std::max_element(diffv.begin(), diffv.end())));
    rect[0] = pts[i_tl];
    rect[1] = pts[i_tr];
    rect[2] = pts[i_br];
    rect[3] = pts[i_bl];
    return rect;
}

static std::pair<Mat, Mat> detect_target(const Mat& frame_bgr) {
    // Returns two masks (for debug) but we’ll actually return contours lists in main
    Mat hsv; cvtColor(frame_bgr, hsv, COLOR_BGR2HSV);

    // Red ranges
    Mat mask1, mask2, maskRed;
    inRange(hsv, Scalar(0,120,70),   Scalar(10,255,255),  mask1);
    inRange(hsv, Scalar(170,120,70), Scalar(180,255,255), mask2);
    maskRed = mask1 | mask2;

    // Blue range
    Mat maskBlue;
    inRange(hsv, Scalar(90,80,160), Scalar(120,255,255), maskBlue);
    return {maskRed, maskBlue};
}

static std::vector<std::vector<Point>> contours_from_mask(const Mat& mask) {
    Mat red_regions; // not strictly needed to find edges; but follow Python’s idea
    // Use mask directly for edges to reduce extra ops
    Mat edges; Canny(mask, edges, 50, 150);

    std::vector<std::vector<Point>> contours;
    findContours(edges, contours, RETR_EXTERNAL, CHAIN_APPROX_SIMPLE);
    std::sort(contours.begin(), contours.end(),
              [](auto& a, auto& b){ return contourArea(a) > contourArea(b); });
    return contours;
}

static std::vector<std::vector<Point2f>> quadrilateral_candidates(const std::vector<std::vector<Point>>& contours,
                                                                  Mat& frame_to_draw,
                                                                  double min_area = 500.0) {
    std::vector<std::vector<Point2f>> quads;
    for (const auto& c : contours) {
        double peri = arcLength(c, true);
        std::vector<Point> approx;
        approxPolyDP(c, approx, 0.02 * peri, true);
        if (approx.size() == 4) {
            double area = contourArea(approx);
            if (area < min_area) continue;
            if (!isContourConvex(approx)) continue;

            std::vector<Point2f> corners;
            corners.reserve(4);
            for (const auto& p : approx) corners.emplace_back((float)p.x, (float)p.y);

            // draw for visualization
            std::vector<std::vector<Point>> drawPoly(1);
            for (auto& q : approx) drawPoly[0].push_back(q);
            drawContours(frame_to_draw, drawPoly, -1, Scalar(0,255,0), 3);

            quads.push_back(corners);
        }
    }
    return quads;
}

// returns "A","B","C","D","E", or empty string if none
static std::string determineLetter(const Mat& markerGray) {
    Mat img_bw;
    threshold(markerGray, img_bw, 200, 255, THRESH_BINARY);
    const uchar white = 255;

    // mimic Python indices
    if (img_bw.at<uchar>(5,5) == white) {
        return ""; // False / not a valid tag
    }

    Mat cropped = img_bw(Rect(25, 25, 125, 125)); // [25:150, 25:150]
    // Access [y,x] in cropped
    auto px = [&](int y, int x)->uchar { return cropped.at<uchar>(y, x); };

    if (px(12, 37) != white && px(12, 62) == white) {
        imshow("Cropped", cropped);
        return "A";
    } else if (px(12, 12) != white && px(12, 37) == white) {
        imshow("Cropped", cropped);
        return "C";
    } else if (px(37, 37) != white && px(12, 112) == white) {
        imshow("Cropped", cropped);
        return "E";
    } else if (px(12, 112) != white && px(62, 112) == white) {
        imshow("Cropped", cropped);
        return "D";
    } else if (px(37, 112) == white && px(62, 122) != white) {
        imshow("Cropped", cropped);
        return "B";
    }
    return "";
}

// Optional: convert rotation matrix to yaw/pitch/roll (deg)
static cv::Vec3d rotToEul(const Mat& R) {
    // R is 3x3 (double)
    double sy = std::sqrt(R.at<double>(0,0)*R.at<double>(0,0) + R.at<double>(1,0)*R.at<double>(1,0));
    bool singular = sy < 1e-6;

    double yaw, pitch, roll;
    if (!singular) {
        yaw   = std::atan2(R.at<double>(2,1), R.at<double>(2,2)); // Z
        pitch = std::atan2(-R.at<double>(2,0), sy);               // Y
        roll  = std::atan2(R.at<double>(1,0), R.at<double>(0,0)); // X
    } else {
        yaw   = std::atan2(-R.at<double>(1,2), R.at<double>(1,1));
        pitch = std::atan2(-R.at<double>(2,0), sy);
        roll  = 0.0;
    }
    return { yaw * 180.0/M_PI, pitch * 180.0/M_PI, roll * 180.0/M_PI };
}

static bool findTranslationAndRotation(const std::vector<Point2f>& image_pts,
                                       cv::Vec3d& tvec, cv::Vec3d& rpy_deg)
{
    std::vector<Point3f> obj = {
        { -fid_size_mm/2.f,  fid_size_mm/2.f, 0.f },
        {  fid_size_mm/2.f,  fid_size_mm/2.f, 0.f },
        {  fid_size_mm/2.f, -fid_size_mm/2.f, 0.f },
        { -fid_size_mm/2.f, -fid_size_mm/2.f, 0.f }
    };
    Mat rvec, tvecm;
    bool ok = solvePnP(obj, image_pts, cameraMatrix, distCoeffs, rvec, tvecm, false, SOLVEPNP_IPPE);
    if (!ok) return false;

    Mat R;
    Rodrigues(rvec, R);           // R: 3x3, double by default
    cv::Vec3d rpy = rotToEul(R);  // yaw, pitch, roll in deg

    tvec = cv::Vec3d(tvecm.at<double>(0,0), tvecm.at<double>(1,0), tvecm.at<double>(2,0));
    rpy_deg = rpy;
    return true;
}

// ----------------- Main -----------------
int main() {
    std::cout.setf(std::ios::fixed); std::cout << std::setprecision(2);

    VideoCapture cap(0);
    if (!cap.isOpened()) {
        std::cerr << "Error: Could not open webcam.\n";
        return 1;
    }

    const int dim = 175;
    std::vector<Point2f> p1 = {
        {0.f, 0.f},
        {(float)(dim-1), 0.f},
        {(float)(dim-1), (float)(dim-1)},
        {0.f, (float)(dim-1)}
    };

    Mat frame;
    while (true) {
        if (!cap.read(frame) || frame.empty()) break;

        // Detect red & blue masks, then contours → quads
        Mat maskRed, maskBlue;
        std::tie(maskRed, maskBlue) = detect_target(frame);

        auto contoursR = contours_from_mask(maskRed);
        auto contoursB = contours_from_mask(maskBlue);

        // For visualization, draw accepted quads on the original frame
        auto quadsR = quadrilateral_candidates(contoursR, frame);
        auto quadsB = quadrilateral_candidates(contoursB, frame);

        // Process RED quads
        for (const auto& corners : quadsR) {
            // Order corners (TL,TR,BR,BL)
            auto ordered = order_corners(corners);

            // Homography and warp to 175x175
            Mat H = findHomography(ordered, p1, RANSAC, 2.0);
            if (H.empty()) continue;

            Mat tag; warpPerspective(frame, tag, H, Size(dim, dim));
            Mat tagGray; cvtColor(tag, tagGray, COLOR_BGR2GRAY);

            std::string letter = determineLetter(tagGray);
            if (letter.empty()) continue;

            cv::Vec3d tvec, rpy_deg;
            if (findTranslationAndRotation(ordered, tvec, rpy_deg)) {
                cout << "Red Tag: " << letter << "\n";
                // Camera position print in their mapping: x=tvec[2], y=tvec[0], z=tvec[1]
                cout << "CAMERA Position x=" << tvec[2] << " y=" << tvec[0] << " z=" << tvec[1] << "\n";
                cout << "Camera tilt -------   yaw=" << rpy_deg[0]
                     << " pitch=" << rpy_deg[1]
                     << " roll="  << rpy_deg[2] << "\n";

                std::string tagmapname = letter + "_RD";
                auto it = TAG_POSES.find(tagmapname);
                if (it != TAG_POSES.end()) {
                    float base_dir = it->second[2];
                    cout << "2D Pose Angle -----   " << (base_dir + (float)rpy_deg[1]) << " degrees\n";

                    float fx = it->second[0];
                    float fy = it->second[1];
                    if (base_dir == 0.f) {
                        cout << "Field Position ----   x=" << (fx - tvec[2]) << " y=" << (fy + tvec[0]) << "\n";
                    } else if (base_dir == 90.f) {
                        cout << "Field Position ----   x=" << (fx - tvec[0]) << " y=" << (fy - tvec[2]) << "\n";
                    } else if (base_dir == 180.f) {
                        cout << "Field Position ----   x=" << (fx + tvec[2]) << " y=" << (fy - tvec[0]) << "\n";
                    } else {
                        cout << "Field Position ----   x=" << (fx + tvec[0]) << " y=" << (fy + tvec[2]) << "\n";
                    }
                }
            }
        }

        // Process BLUE quads
        for (const auto& corners : quadsB) {
            auto ordered = order_corners(corners);
            Mat H = findHomography(ordered, p1, RANSAC, 2.0);
            if (H.empty()) continue;

            Mat tag; warpPerspective(frame, tag, Size(dim, dim), H);
            Mat tagGray; cvtColor(tag, tagGray, COLOR_BGR2GRAY);

            std::string letter = determineLetter(tagGray);
            if (letter.empty()) continue;

            cv::Vec3d tvec, rpy_deg;
            if (findTranslationAndRotation(ordered, tvec, rpy_deg)) {
                cout << "Blue Tag: " << letter << "\n";
                cout << "CAMERA Position x=" << tvec[2] << " y=" << tvec[0] << " z=" << tvec[1] << "\n";
                cout << "Camera tilt -------   yaw=" << rpy_deg[0]
                     << " pitch=" << rpy_deg[1]
                     << " roll="  << rpy_deg[2] << "\n";

                std::string tagmapname = letter + "_BL";
                auto it = TAG_POSES.find(tagmapname);
                if (it != TAG_POSES.end()) {
                    float base_dir = it->second[2];
                    cout << "2D Pose Angle -----   " << (base_dir + (float)rpy_deg[1]) << " degrees\n";

                    float fx = it->second[0];
                    float fy = it->second[1];
                    if (base_dir == 0.f) {
                        cout << "Field Position ----   x=" << (fx - tvec[2]) << " y=" << (fy + tvec[0]) << "\n";
                    } else if (base_dir == 90.f) {
                        cout << "Field Position ----   x=" << (fx - tvec[0]) << " y=" << (fy - tvec[2]) << "\n";
                    } else if (base_dir == 180.f) {
                        cout << "Field Position ----   x=" << (fx + tvec[2]) << " y=" << (fy - tvec[0]) << "\n";
                    } else {
                        cout << "Field Position ----   x=" << (fx + tvec[0]) << " y=" << (fy + tvec[2]) << "\n";
                    }
                }
            }
        }

        imshow("Webcam Detection", frame);
        int k = waitKey(1);
        if ((k & 0xFF) == 'q') break;
    }

    cap.release();
    destroyAllWindows();
    return 0;
}

